add ipam document (#233)
cybozu-go/coil: docs/design.md

For example, if an address pool defines that the size of an address block is 2<sup>5</sup>, `coil-controller` will curve an address block for IPv4 with `/27` subnet mask out of the pool, and assigns it to a node.

> When assigning an address, if `coil-controller` assigns the last released address again, it is possible to use old routes and it is not possible to establish a new connection via egress NAT.

Reviewer: > it is not possible to establish a new connection via egress NAT.

I guess this is due to Cilium, not Coil.

> Therefore, avoiding reassigning the last released address, it records the index of the last assigned address.

Reviewer: I'd say something like, "In general, avoiding immediate reuse of IP addresses is better to not confuse other software or peripherals.".

And in the next time, it assigns an address next to the recorded index.

> When we use an address block with the size `/32`, it may occur that reassignment of the same address block.

Reviewer: ```suggestion
When we use an address block with the size `/32`, there is a high chance of reusing the same address immediately.
```

> However, the address block with the size `/32` is usually used as a global address.

Reviewer: ```suggestion
However, the address block with the size `/32` is usually used for global addresses.
```

A global address will not frequently allocate and release.

> Thus, we don't implement this.

Reviewer: But if it happens, does Cilium work well?

Author: In this situation, I think Cilium cannot handle the connection well.
But the frequency may be less than the case of reusing addresses.

As a result of the discussion, if we just address the issue we face now, considering the cost and time of fix, we don't need to deal with address blocks case.

Should we deal with in this fix?

Reviewer: I'm not against you and the team, but I'm a bit concerned that the loss of connectivity would be more severe for global addresses rather than cluster addresses.

Author: Now I think it is enough that we are aware of this problem and we mention it in this document.
In the long term, I also think that it is better to fix this.

Reviewer: ok

For example, if an address pool defines that the size of an address block is 2<sup>5</sup>, `coil-controller` will curve an address block for IPv4 with `/27` subnet mask out of the pool, and assigns it to a node.

In general, avoiding immediate reuse of IP addresses is better not to confuse other software or components.
To avoid such immediate reuse, `coil-controller` remembers the last used address, and it assigns the address from the next address.

The same problem may occur when we use address blocks of the size `/32`.
In this case, there is a high chance of reusing the same address immediately.
However, the address block of the size `/32` is usually used for public addresses.

> Public IP addresses are not allocate and release frequently.

Reviewer 2: ```suggestion
Public IP addresses are not allocated and released frequently.
```

For example, if an address pool defines that the size of an address block is 2<sup>5</sup>, `coil-controller` will curve an address block for IPv4 with `/27` subnet mask out of the pool, and assigns it to a node.

> When assigning an address, if `coil-controller` reuses the last released address immediately, it tends to use old routes.

Reviewer: > When assigning an address, if `coil-controller` reuses the last released address immediately, it tends to use old routes.

I don't get what this sentence means. Why "it tends to use old routes"??
I think this paragraph can make sense without this sentence.

Author: I wanted to show some examples of concerns of reusing addresses.
But I certainly may not need it.

In general, avoiding immediate reuse of IP addresses is better not to confuse other software or components.

> Avoiding this, `coil-controller` remembers the last used address, and it assigns the address from the next address.

Reviewer: > Avoiding this,

This phrase is ambiguous. How about "To avoid such immediate reuse,"?

> Same problem may occurs when we use address blocks of the size `/32`.

Reviewer: ```suggestion
The same problem may occur when we use address blocks of the size `/32`.
```
